libcob/profiling.c

```diff
@@ -0,0 +1,464 @@
+   License, or (at your option) any later version.
+
+   GnuCOBOL is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with GnuCOBOL.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include <limits.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+
+/* include internal and external libcob definitions, forcing exports */
+#define COB_LIB_EXPIMP
+#include "coblocal.h"
+
+#include "tarstamp.h"
+#include "config.h"
+#include "common.h"
+
+#include <sys/types.h>
+#ifdef HAVE_UNISTD_H
+#include <unistd.h>
+#endif
+
+#ifdef	_WIN32
+#define WIN32_LEAN_AND_MEAN
+#include <windows.h>
+#undef MOUSE_MOVED
+#include <process.h>
+#include <io.h>
+#include <fcntl.h>	/* for _O_BINARY only */
+#endif
+
+/* Local types and variables */
+
+struct cob_prof_module_list {
+	struct cob_prof_module *info ;
+	struct cob_prof_module_list *next;
+};
+
+static struct cob_prof_module_list *prof_info_list ;
+
+/* We maintain a stack of the procedures entered as 3 different
+ * arrays, with "current_idx" being the stack pointer. */
+static cob_ns_time              *start_times;
+static int                      *called_procedures;
+static struct cob_prof_module*  *called_runtimes;
+/* Current size of previous arrays */
+static int max_prof_depth;
+static int current_idx = -1;
+
+/* Whether profiling is active or not. */
+static int is_active = 0;
+/* Whether we are in testsuite mode */
+static int is_test = 0;
+
+/* Remember static and dynamic configuration */
+static cob_global               *cobglobptr = NULL;
+static cob_settings             *cobsetptr = NULL;
+
+
+
+/* Return the current time in nanoseconds. The result is guarranteed
+ * to be monotonic, by using an internal storage of the previous
+ * time. */
+static cob_ns_time
+get_ns_time (void)
+{
+	static cob_ns_time ns_time = 0;
+	unsigned long long nanoseconds;
+
+	if (is_test){
+		ns_time += 10000;
+		return ns_time;
+	} else {
+#if defined (_WIN32)
+		LARGE_INTEGER performance_counter;
+		LARGE_INTEGER performance_frequency;
+#elif defined (HAVE_CLOCK_GETTIME)
+		struct timespec ts;
+#else
+		clock_t c;
+#endif
+
+#if defined (_WIN32)
+		QueryPerformanceCounter(&performance_counter);
+		QueryPerformanceFrequency(&performance_frequency);
+		performance_counter.QuadPart *= 1000000000;
+		performance_counter.QuadPart /= performance_frequency.QuadPart;
+		nanoseconds = performance_counter.QuadPart;
+#elif defined (HAVE_CLOCK_GETTIME)
+		clock_gettime(CLOCK_MONOTONIC, &ts);
+		nanoseconds = ts.tv_sec * 1000000000 + ts.tv_nsec;
+#else
+		c = clock();
+		nanoseconds = c * 1000000000;
+		nanoseconds /= CLOCKS_PER_SEC;
+#endif
+	}
+	if (nanoseconds>ns_time) ns_time = nanoseconds;
+	return ns_time;
+}
```

Comment:
Is this still an issue ?

Reply:
yes (and additional, per last review: please have the cob_ns_time only static in the testsuite block); the `CLOCK_MONOTONIC` is _not_ a given (neither the define, nor the working return), so that must be adjusted (I _think_ my suggestion above would work fine).